platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/services/DefaultIsolatedProjectEvaluationListenerProvider.kt

```diff
@@ -73,22 +84,34 @@ class DefaultIsolatedProjectEvaluationListenerProvider(
         } ?: action
 
     override fun isolateFor(gradle: Gradle): ProjectEvaluationListener? = when {
-        beforeProject.isEmpty() && afterProject.isEmpty() -> null
+        beforeProject.isEmpty() && afterProject.isEmpty() && allprojects.isEmpty() -> null
         else -> {
-            val isolate = isolate(
-                IsolatedProjectActions(beforeProject, afterProject),
-                IsolateOwners.OwnerGradle(gradle)
-            )
-            clear()
```

Comment:
Because now `clear()` implies assignment `allprojectsAction=null` alongside with cleaning of the registered callbacks, what is not desirable at the moment of the invocation.

Obviously, this is easily managable, I just don't have any strong preference about this early cleanup of the callbacks.

Reply:
Similar to how we modeled the whole isolated actions state to release isolated callbacks that have already been executed, we should release resources here as soon as they are no longer needed.